Fix possible BUGCHECK "cannot find tip page" during validation with -full or -mend option

ilya071294 opened the pull request:
If a primary record version has a transaction number which is greater than NT and seems corrupted, its state will be treated as tra_active to avoid an attempt to fetch a non-existing TIP page.

ilya071294 committed:
Fix possible BUGCHECK "cannot find tip page" during validation with -full or -mend option

If a primary record version has a transaction number which is greater than NT and seems corrupted, its state will be treated as tra_active to avoid an attempt to fetch a non-existing TIP page.

hvlad reviewed (approved):
I have no objection, just one question - what consequences you expect with such change, if it is `Next` on disk is wrong (less than actual value) ? 
Perhaps such condition could be detected in `walk_tip()` ?

ilya071294 commented:
> if it is `Next` on disk is wrong (less than actual value)

The actual value - do you mean it as the number of the last non-active transaction on the TIP?

hvlad commented:
> > if it is `Next` on disk is wrong (less than actual value)
> 
> The actual value - do you mean it as the number of the last non-active transaction on the TIP?

Yes, exactly.

ilya071294 commented:
> Yes, exactly.

If such corruption is detected, is it possible to fix it safely? Which page can we trust more - the header or the TIP?

hvlad commented:
> If such corruption is detected, is it possible to fix it safely? Which page can we trust more - the header or the TIP?

If `Next` points to the last TIP page and if TIP chain is valid, then both header and last TIP could be trusted in the same degree, IMHO.

dyemanov commented:
Do we expect any more commits inside this PR? Or can it be merged now?

ilya071294 committed:
1. Detect a corruption where the TIP has a transaction with a non-zero state which number is greater than Next transaction. When -mend option is specified, fix it by advancing Next transaction.
2. Do not call TRA_extend_tip() when it's possible to reconstruct the TIP chain using tip_next from the previous page. Also TRA_extend_tip() call is only allowed when the sequence is not found at the end of the vector. It prevents the potential loss of TIPs because this function unconditionally allocates a new TIP and overwrites tip_next.
3. Check the count of elements in the vector before accessing them to avoid an assertion failure in getElement().

ilya071294 committed:
Merge branch 'master' into fb_master_gfix_tip_fetch_fix

ilya071294 committed:
Merge branch 'master' into fb_master_gfix_tip_fetch_fix

ilya071294 commented:
Conflicts are solved. Please review.
